File: cpp/board_game_permutations_utils.cpp
#include "board_game_permutations_utils.h"

long long get_current_time_ms() {
  std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
  std::chrono::system_clock::duration duration = now.time_since_epoch();
  std::chrono::milliseconds millis =
      std::chrono::duration_cast<std::chrono::milliseconds>(duration);
  return millis.count();
}

void print_board(const Board &board) {
  std::cout << "Board:\n";
  for (const std::vector row : board) {
    for (int val : row) {
      std::string str_int = std::to_string(val);
      std::string padding(4 - str_int.length(), ' ');
      std::cout << val << padding;
    }
    std::cout << '\n';
  }
  std::cout << std::endl;
}

Board rotate_board_90(Board board) {
  int size = board.size();

  // transpose
  for (int i = 0; i < size; ++i) {
    for (int j = i + 1; j < size; ++j) {
      std::swap(board[i][j], board[j][i]);
    }
  }

  // reverse each row
  for (int i = 0; i < size; ++i) {
    std::reverse(board[i].begin(), board[i].end());
  }

  return board;
}

Board generate_initial_board(const int &size) {
  return Board(size, std::vector<int>(size, 0));
}

std::string hash_board(const Board &board) {
  int size = board.size();
  std::string board_str;
  board_str.reserve(board.size() * board.size() * 3);

  for (int r = 0; r < size; ++r) {
    for (int c = 0; c < size; ++c) {
      // board_str += std::to_string(r) + ","+ std::to_string(c) + "," +
      board_str += std::to_string(board[r][c]) + "|";
    }
  }

  return board_str;
}

Board board_hash_to_array(const std::string &board_hash_str, int size) {
  std::vector<int> flat_values;
  std::stringstream ss(board_hash_str);
  std::string token;

  while (std::getline(ss, token, '|')) {
    flat_values.push_back(std::stoi(token));
    // } catch (const std::invalid_argument &e) {
    //   throw std::runtime_error("Invalid number in board hash: " + token);
    // }
  }

  // if (flat_values.size() != size * size) {
  //   throw std::runtime_error(
  //       "Hash does not contain the right number of values for the board
  //       size.");
  // }

  // populate 2D board

  Board board(size, std::vector<int>(size));
  for (int i = 0; i < size * size; ++i) {
    int row = i / size;
    int col = i % size;
    board[row][col] = flat_values[i];
  }

  return board;
}

std::vector<std::vector<Board>> split_list(const std::vector<Board> &list,
                                           int n) {
  int total = list.size();
  int split = std::max(total / n, 1);

  std::vector<std::vector<Board>> splits;

  for (int i = 0; i < n; ++i) {
    int start = i + (i * split);
    int end = i + ((i + 1) * split) + 1;

    end = std::min(end, total);

    if (start >= total)
      break;

    splits.emplace_back(list.begin() + start, list.begin() + end);
  }

  return splits;
}

std::string write_to_file(const std::unordered_set<std::string> &results,
                          const std::string &folder_name) {
  std::string filename =
      folder_name + "_" + std::to_string(get_current_time_ms()) + ".txt";
  fs::path filepath = fs::current_path() / folder_name / filename;

  std::ofstream file(filepath);
  for (const std::string &result : results) {
    file << result << "\n";
  }

  return filename;
}

std::string write_to_file(const std::vector<Board> &new_increments,
                          const std::string &folder_name) {
  std::string filename =
      folder_name + "_" + std::to_string(get_current_time_ms()) + ".txt";
  fs::path filepath = fs::current_path() / folder_name / filename;

  std::ofstream file(filepath);
  for (const Board &new_increment : new_increments) {
    file << hash_board(new_increment) << "\n";
  }

  return filename;
}

std::vector<bool> check_results_vs_files(
    const std::vector<std::pair<Board, std::vector<std::string>>> &results_list,
    const std::vector<std::string> &result_files) {
  std::vector<bool> is_new_check(results_list.size(), true);

  for (const std::string &filename : result_files) {
    fs::path filepath = fs::current_path() / "results" / filename;
    std::ifstream file(filepath);

    std::unordered_set<std::string> temp_results;
    temp_results.reserve(10'000'000);
    std::string line;
    while (std::getline(file, line)) {
      temp_results.insert(line);
    }

    for (size_t i = 0; i < results_list.size(); ++i) {
      if (!is_new_check[i])
        continue;

      const std::vector<std::string> &rotated_hashes = results_list[i].second;
      for (const std::string &h : rotated_hashes) {
        if (temp_results.count(h) > 0) {
          is_new_check[i] = false;
        }
      }
    }
  }

  return is_new_check;
}

long long get_file_item_count(std::vector<std::string> files,
                              std::string folder_name) {
  long long count = 0;

  for (size_t i = 0; i < files.size(); ++i) {
    std::filesystem::path file_path =
        std::filesystem::current_path() / folder_name / files[i];
    std::ifstream file(file_path);

    std::string line;
    while (std::getline(file, line)) {
      ++count;
    }

    file.close();
  }

  return count;
}

void display_round_stats(int round, int rounds, std::time_t start,
                         std::time_t round_start, int new_board_count,
                         int result_count) {
  long long now = get_current_time_ms();

  double round_seconds_elapsed = (now - round_start) / 1000.0;
  double total_seconds_elapsed = (now - start) / 1000.0;

  double round_hours_elapsed = round_seconds_elapsed / 3600.0;
  double total_hours_elapsed = total_seconds_elapsed / 3600.0;

  double percent_done = (round + 0.0) / rounds * 100.0;
  double time_spent_ratio =
      ((round_seconds_elapsed + 0.0) / (total_seconds_elapsed + 1e-6)) * 100.0;

  std::cout << std::fixed << std::setprecision(1);

  std::cout << "Round " << round << "\t\t" << round << "/" << rounds << "\t";
  std::cout << percent_done << "%" << std::endl;

  std::cout << "\t" << new_board_count << " unique boards found this round";
  std::cout << std::endl;

  std::cout << "\t" << result_count << " unique boards found so far";
  std::cout << std::endl;

  std::cout << "\t" << round_hours_elapsed << " hours or ";
  std::cout << round_seconds_elapsed << " seconds elapsed this round";
  std::cout << std::endl;

  std::cout << "\t" << total_hours_elapsed << " hours or ";
  std::cout << total_seconds_elapsed << " seconds elapsed so far";
  std::cout << std::endl;

  std::cout << "\t" << time_spent_ratio;
  std::cout << "% of total time spent on this round";
  std::cout << std::endl;
  std::cout << std::endl;
}

File: python/board_game_permutations_utils.py
import time
from pathlib import Path
import multiprocessing as mp
from multiprocessing import shared_memory
from concurrent.futures import ThreadPoolExecutor
import numpy as np

def generate_initial_board(size):
    initial_board = np.zeros((size, size))
    return initial_board


def hash_board(board):
    size = len(board)
    board_str = ""
    for r in range(size):
        for c in range(size):
            # board_str += f"{r},{c},{board[r, c]}|"
            board_str += f"{board[r, c]}|"

    return board_str


def board_hash_to_array(board_hash_str, size):
    board = np.zeros((size, size))
    squares = board_hash_str[:-1].split("|")
    for i, square in enumerate(squares):
        # r, c, v = square.split(",")
        r = i / size
        c = i % size
        board[int(r), int(c)] = int(float(square))

    return board


def split_list(list_, n):
    split = max(int(len(list_) / n), 1)
    splits = []
    for i in range(n):
        splits.append(list_[i + (i * split) : i + ((i + 1) * split) + 1])

    return splits


def write_to_file(items, folder_name):
    filename = f"{folder_name}_{int(time.time() * 1e3)}.txt"
    with open(Path.cwd() / folder_name / filename, "w") as file:
        if folder_name == "results":
            file.write("\n".join(items))
        elif folder_name == "new_increments":
            file.write("\n".join(hash_board(item) for item in items))
        else:
            pass

    return filename




def check_results_vs_files(result_files, results_lists, n_jobs):

    def check_results_vs_file_mp(results_list, is_new_check):
        for i, (board_increment, min_board_hash) in enumerate(results_list):
            if is_new_check[i] and min_board_hash in temp_results:
                is_new_check[i] = False

        return is_new_check

    is_new_checks = [[True] * len(l) for l in results_lists]

    for result_file in result_files:
        with open(Path.cwd() / "results" / result_file, "r") as file:
            temp_results = set(file.read().splitlines())

        with ThreadPoolExecutor(max_workers=n_jobs) as executor:
            futures = [executor.submit(check_results_vs_file_mp, r, isc) for r, isc in zip(results_lists, is_new_checks)]
            is_new_checks = [f.result() for f in futures]

    return is_new_checks

def get_file_item_count(files, folder_name):
    if folder_name == "new_increments":
        count = 0
        for filename in files:
            with open(Path.cwd() / folder_name / filename, "r") as file:
                temp = file.read().splitlines()
            count += len(temp)

    elif folder_name == "results":
        count_per_file = 0
        if files:
            with open(Path.cwd() / folder_name / files[0], "r") as file:
                temp = file.read().splitlines()
            count_per_file += len(temp)
        count = len(files) * count_per_file

    return count


def display_round_stats(round_, rounds, start, round_start, new_board_count, result_count):
    now = time.time()

    round_seconds_elapsed = round(now - round_start, 1)
    round_hours_elapsed = round(round_seconds_elapsed / 3600, 2)

    seconds_elapsed = round(now - start, 1)
    hours_elapsed = round(seconds_elapsed / 3600, 2)

    print(f"Round {round_}\t\t{round_}/{rounds}\t{round(round_ / rounds * 100, 1)}%")
    print(f"\t{new_board_count} unique boards found this round")
    print(f"\t{result_count} unique boards found so far")
    print(f"\t{round_hours_elapsed} hours or {round_seconds_elapsed} seconds elapsed this round")
    print(f"\t{hours_elapsed} hours or {seconds_elapsed} seconds elapsed so far")
    print(f"\t{round((round_seconds_elapsed / (seconds_elapsed + 1e-6)) * 100, 1)}% of total time spent on this round")
    print()
